### PR TITLE
Refactor Cohere Model

### PR DESCRIPTION
**Pull request by saurabhdash2512**
# What does this PR do?

Refactor Cohere Model

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

 @ArthurZucker and @younesbelkada

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/cohere/modeling_cohere.py [hunk 1]
```diff
@@ -76,10 +76,9 @@ def _get_unpad_data(attention_mask):
 
 
 class CohereLayerNorm(nn.Module):
-    def __init__(self, hidden_size, eps=1e-5, bias=False):
+    def __init__(self, param_shape=None, eps=1e-5, bias=False):
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:79]:
```suggestion
    def __init__(self, hidden_size =None, eps=1e-5, bias=False):
        """ The hidden size can be a tuple or an int. If a tuple is used, the layer will be applied on both dim """
```

File: src/transformers/models/cohere/modeling_cohere.py [hunk 2]
```diff
@@ -89,8 +88,6 @@ def forward(self, hidden_states):
         variance = (hidden_states - mean).pow(2).mean(-1, keepdim=True)
         hidden_states = (hidden_states - mean) * torch.rsqrt(variance + self.variance_epsilon)
         hidden_states = self.weight.to(torch.float32) * hidden_states
-        if self.bias is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:92]: I am guessing this was never used?

Comment (ArthurZucker) [reply src/transformers/models/cohere/modeling_cohere.py:92]: This layer norm falls back to being `# Copied from Llama`

Comment (ArthurZucker) [reply src/transformers/models/cohere/modeling_cohere.py:92]: Also missing on the `def rotate_half(x):` function

File: src/transformers/models/cohere/modeling_cohere.py [hunk 3]
```diff
@@ -154,11 +150,14 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
...
     """
+    dtype = q.dtype
+    q = q.float()
+    k = k.float()
     cos = cos.unsqueeze(unsqueeze_dim)
     sin = sin.unsqueeze(unsqueeze_dim)
     q_embed = (q * cos) + (rotate_half(q) * sin)
     k_embed = (k * cos) + (rotate_half(k) * sin)
-    return q_embed, k_embed
+    return q_embed.to(dtype=dtype), k_embed.to(dtype=dtype)
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:160]: if this is done outside `apply_rotary_pos_emb` we can keep the copied from 😉 but it's not an issue

File: src/transformers/models/cohere/modeling_cohere.py [hunk 4]
```diff
@@ -216,13 +214,21 @@ def __init__(self, config: CohereConfig, layer_idx: Optional[int] = None):
...
+        self.use_qk_norm = config.use_qk_norm
 
         if (self.head_dim * self.num_heads) != self.hidden_size:
             raise ValueError(
                 f"hidden_size must be divisible by num_heads (got `hidden_size`: {self.hidden_size}"
                 f" and `num_heads`: {self.num_heads})."
             )
 
+        if self.use_qk_norm:
+            # When sharding the model using Tensor Parallelism, need to be careful to use n_local_heads
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:226]: Not sure if this comment is relevant as the model is not sharded by default

**Review by ArthurZucker — COMMENTED** [review]
Thanks. 
Let's keep BC for the layer norm name. 
While we are at it, missing copied from! 🤗

File: src/transformers/models/cohere/modeling_cohere.py (hunk 4 above)
Comment (saurabhdash2512) [reply src/transformers/models/cohere/modeling_cohere.py:226]: Oh this is to warn others who port this to other frameworks from HF.

File: src/transformers/models/cohere/modeling_cohere.py (hunk 2 above)
Comment (saurabhdash2512) [reply src/transformers/models/cohere/modeling_cohere.py:92]: This is different from Llama's Norm with is RMS and at different precision.

Comment (ArthurZucker) [reply src/transformers/models/cohere/modeling_cohere.py:92]: Yep sorry

**Review by ArthurZucker — APPROVED** [review]

**Comment by ArthurZucker** [comment]
As a followup, adding integration tests for the new model will be nice.

**Comment by fblissjr** [comment]
I think there's an issue with the tokenizer.json (or the way transformers is converting). Random multilingual out of nowhere.

See: https://huggingface.co/CohereForAI/c4ai-command-r-plus/discussions/15

Found this after converting to mlx-lm, but was found to be reproducible via the following:

You can reproduce with:

```from transformers import AutoTokenizer
 
tokenizer = AutoTokenizer.from_pretrained("CohereForAI/c4ai-command-r-plus")
tokenizer.save_pretrained(".")```

**Comment by awni** [comment]
Concretely, loading -> saving -> loading the tokenizer all through HF `AutoTokenizer` produces a new tokenizer than the initial loaded one with worse behavior. Presumably that should not be the case.

**Comment by fblissjr** [comment]
worth noting - the bitsandbytes 4 bit repo linked from the main cohere repo is also a different size and looks very different from the original model repo's tokenizer.json (https://huggingface.co/CohereForAI/c4ai-command-r-plus-4bit/blob/main/tokenizer.json).

**Comment by fblissjr** [comment]
another interesting difference between the 4 bit bnb tokenizer and the original - in the original one, token id token id 255001 <|END_OF_TURN_TOKEN|>, special is set to False. In the 4bit bnb one, it's True.

**Comment by ArthurZucker** [comment]
I have no idea what you are talking about? A 4bit tokenizer? could you actually open an issue with a repro and the issue?

**Comment by ahmetustun** [comment]
I think the difference between two tokenizers.json is unicode encoding after `tokenizer.save_pretrained` which is used to save tokenizer in 4bit model. Also, in the config <|END_OF_TURN_TOKEN|> token is also set as special because it is used as eos_token, which is also overwritten in the original tokenizer (command-r-plus) as well. Therefore, two tokenizer should work exactly the same.

**Comment by fblissjr** [comment]
> I have no idea what you are talking about? A 4bit tokenizer? could you actually open an issue with a repro and the issue?

What @ahmetustun mentioned is the difference - not a 4 bit tokenizer, the tokenizer.json in https://huggingface.co/CohereForAI/c4ai-command-r-plus vs. https://huggingface.co/CohereForAI/c4ai-command-r-plus-4bit, which @ahmetustun clarified.

Not at my workstation now, but if nobody else is seeing this issue, I'll assume I've got something wrong on my end. Thanks for the clarification.

**Comment by ahmetustun** [comment]
Please left further comment in the model repo if the problem continues. Thanks @fblissjr.

**Pull request merged**